File: backend/app/engines/context/ContextResult.py
from __future__ import annotations

from typing import Literal
from typing import TypedDict


class ContextLiquidityTaken(TypedDict):
    timeframe: Literal["H4"]
    valid: bool
    side: Literal["BSL", "SSL", "NONE"]
    level: float | None


class ContextOrderBlock(TypedDict):
    timeframe: Literal["H1"]
    valid: bool
    blockType: Literal["BULLISH", "BEARISH", "NONE"]
    low: float | None
    high: float | None


class ContextFairValueGap(TypedDict):
    timeframe: Literal["H1"]
    valid: bool
    gapType: Literal["BULLISH", "BEARISH", "NONE"]
    low: float | None
    high: float | None


class ContextResult(TypedDict):
    valid: bool
    contextScore: int
    reasons: list[str]
    warnings: list[str]
    liquidityTaken: ContextLiquidityTaken
    orderBlock: ContextOrderBlock
    fairValueGap: ContextFairValueGap

File: frontend/src/models/ContextResult.ts
export type LiquiditySide = "BSL" | "SSL" | "NONE";

export type ContextBlockType = "BULLISH" | "BEARISH" | "NONE";

export interface ContextLiquidityTaken {
  timeframe: "H4";
  valid: boolean;
  side: LiquiditySide;
  level: number | null;
}

export interface ContextOrderBlock {
  timeframe: "H1";
  valid: boolean;
  blockType: ContextBlockType;
  low: number | null;
  high: number | null;
}

export interface ContextFairValueGap {
  timeframe: "H1";
  valid: boolean;
  gapType: ContextBlockType;
  low: number | null;
  high: number | null;
}

export interface ContextResult {
  valid: boolean;
  contextScore: number;
  reasons: string[];
  warnings: string[];
  liquidityTaken: ContextLiquidityTaken;
  orderBlock: ContextOrderBlock;
  fairValueGap: ContextFairValueGap;
}
